import type { NextFunction, Request, Response } from "express";
import bcrypt from "bcryptjs";
import { registerUser } from "./userControllers";
import type { UserStructure } from "../../database/models/User";
import User from "../../database/models/User";
import CustomError from "../../CustomError/CustomError";
import mongoose from "mongoose";

beforeEach(() => {
  jest.clearAllMocks();
});

const res: Partial<Response> = {
  status: jest.fn().mockReturnThis(),
  json: jest.fn(),
};

const next = jest.fn();

describe("Given a register controller", () => {
  const user: UserStructure = {
    password: "1234",
    email: "mireia@gmail.com",
    name: "mireia",
    gender: "f",
    level: 2,
  };

  const req: Partial<Request> = {
    body: user,
  };

  describe("When it receives an email 'mireia@gmail.com' and a password '1234''", () => {
    test("Then it should invoke its method status with 201 and its method json with the user id and the email", async () => {
      const expectedStatus = 201;
      const hashedPassword = "hashedpassword";
      req.body = user;
      const id = new mongoose.Types.ObjectId();

      bcrypt.hash = jest.fn().mockResolvedValue(hashedPassword);

      User.create = jest.fn().mockResolvedValue({
        email: user.email,
        password: hashedPassword,
        name: user.name,
        gender: user.gender,
        level: user.level,
        _id: id,
      });

      await registerUser(req as Request, res as Response, next as NextFunction);

      expect(res.status).toHaveBeenCalledWith(expectedStatus);
      expect(res.json).toHaveBeenCalledWith({
        id,
        email: user.email,
        name: user.name,
        gender: user.gender,
        level: user.level,
      });
    });
  });

  describe("When it receives an email 'mireia@gmail.com' that already exists", () => {
    test("Then it should next with status 409 and a public message 'User already registered'", async () => {
      const error = new CustomError(
        "duplicate key",
        409,
        "User already registered"
      );

      User.create = jest.fn().mockRejectedValueOnce(error);

      await registerUser(req as Request, res as Response, next as NextFunction);

      expect(next).toHaveBeenCalledWith(error);
    });
  });

  describe("When it receives a request with email 'mireia@gmail.com', password '12345678' and bcrypt rejects", () => {
    test("Then next should be invoked with the error generated by bcrypt", async () => {
      req.body = user;
      const bcryptError = new Error();
      bcrypt.hash = jest.fn().mockRejectedValue(bcryptError);

      await registerUser(req as Request, null, next as NextFunction);

      expect(next).toHaveBeenCalledWith(bcryptError);
    });
  });
});
